include/functional/fail.hpp

```diff
@@ -15,38 +15,42 @@
 
 namespace fn {
 
-struct fail_t final {
-  auto operator()(auto &&fn) const noexcept
-      -> functor<fail_t, std::decay_t<decltype(fn)>>
+constexpr static struct fail_t final {
+  auto operator()(auto &&fn) const noexcept -> functor<fail_t, decltype(fn)>
   {
-    using functor_type = functor<fail_t, std::decay_t<decltype(fn)>>;
-    return functor_type{{std::forward<decltype(fn)>(fn)}};
+    return {std::forward<decltype(fn)>(fn)};
   }
-};
-constexpr static fail_t fail = {};
+} fail = {};
 
-auto monadic_apply(some_optional auto &&v, fail_t const &,
-                   some_tuple auto &&fn) noexcept -> auto
-  requires(std::tuple_size_v<std::decay_t<decltype(fn)>> == 1)
+auto monadic_apply(some_expected auto &&v, fail_t, auto &&fn) noexcept
```

Comment:
For `auto &&fn`, could we not use `std::invocable` concept instead of the custom `requires requires`?

Reply:
will do